Test: forced timeout dispute

daiagi opened the pull request:
- Test for forced timeout dispute
- Minor bug fix: avoid scheduling tasks when state manager is disposed
- Bug fix: `onChainTimestamp`  was not being set on the storage `Block`
    * removed the "band-aid" code that was superficially fixing this bug ([link](https://github.com/peer3to/state-channels-plus/commit/3cbe64ea1a1cd1583ccdd9b26c697c02609a5c27#r169464660))
- This and that cleaner API's on `PeerTestHarness` and of-course, added the ability to submit "junk" as block calldata

daiagi committed:
minor bug fix: avoid scheduleTask if StateManager is disposed

daiagi committed:
log `isForced` in onDisputeCommitted event handler

daiagi committed:
test: isForced timeout

daiagi committed:
bug fix: update storage block onChainTimestamp

lukastanisic99 commented on src/stateManager/StateManager.ts:
```diff
@@ -1308,22 +1313,6 @@ class StateManager {
-                        previousBlock.height,
-                        previousBlock.author
-                    )
-                ).found;
```
why is this removed

daiagi replied:
see issue description:
> removed the "band-aid" code that was superficially fixing this bug ([link](https://github.com/peer3to/state-channels-plus/commit/3cbe64ea1a1cd1583ccdd9b26c697c02609a5c27#r169464660))

lukastanisic99 replied:
`tryTimeoutParticipant` should take into account the `previousBlock.relevantTimestamp` - that means that pipeline has run the block and if everything is ok it persisted the on-chain timestamp that will be taken into account for the calculation and if the timestamp is not persisted what's posted on-chain is mostly junk, `but it's still posted and persisted on-chain`, so regardless we'd have to set `previousBlockProducerPostedCalldata = true` (even though we might timeout early since it's junk), but it has to be set as `true` so we signal on-chain `I'm aware that calldata for this block exists` when making the on-chain transaction - this is a race condition guard that protects the sender of the transaction from potential race condition - so.... if junk is persisted and this is set to false the transaction will revert since there's a mismatch between the expectation and reality 

This code here just takes into account if the on-chain timestamp is persisted in storage, which means it's accepted - it doesn't cover the junk case (not in storage, but available on-chain)

daiagi replied:
honestly i am not sure what you are talking about
this is a new code that i added in the commit i asked you to review in the other PR
https://github.com/peer3to/state-channels-plus/pull/204#issuecomment-3472306867
it was just masking the issue that onChainTimestamp was undefined

since i have fixed it from the root cause, i have removed this code.
not see what the issue is.
original code, that you have written is

```ts
           previousBlockProducerPostedCalldata: previousBlock
                ? Boolean(previousBlock.onChainTimestamp)
                : false,
```

lukastanisic99 replied:
Yes I know and when I saw your `aid implementation` and made the [comment](https://github.com/peer3to/state-channels-plus/commit/3cbe64ea1a1cd1583ccdd9b26c697c02609a5c27#r169464660) I realized I forgot to do the check above. 

Test for this case:
- Produce and gossip a correct block N and post some junk as calldata for block N on-chain, 
- The next peer doesn't produce block N+1
- A timeout will be initiated for block N+1 after some time

What should happen:
- previousBlock.onChainTimestamp will be undefined (pipeline didn't accept junk) -> relevantTimestamp == block.timestamp; Temeout will initiate after p2p+agreement+fallback time (which is good). `previousBlockProducerPostedCalldata` should be set to `true` matching the expectation that the storage slot that corresponds to the previousBlock on-chain storage slot is set. The timeout goes through and is accepted. 

What will happen under: 
```tyepscript
previousBlockProducerPostedCalldata: previousBlock
                ? Boolean(previousBlock.onChainTimestamp)
                : false,
```
`previousBlockProducerPostedCalldata` will be `false` even though the storage slot on-chain is set, which will cause the transaction to revert - https://github.com/peer3to/state-channels-plus/blob/dac95097d2a321488f187540b1d36607b219eaa1/contracts/V1/StateChannelDiamondProxy/DisputeManagerFacet.sol#L124

daiagi replied:
restored and added a test
however, this triggered a new , yet untested, buggy behavior:
- Produce and gossip a correct block N and post some junk as calldata for block N on-chain, 
    * The posting of junk calldata triggers the other peers to create a dispute
    * The uploading of this dispute fails for unknown reason
- The next peer doesn't produce block N+1
- A timeout will be initiated for block N+1 after some time
    * The failure of the dispute above, corrupts something (idk what yet) that causes the timeout dispute to not be created

daiagi replied:
see comments in the `"should create forced timeout when peer posts junk calldata that is rejected"` test
and the error in the gh workflow [log](https://github.com/peer3to/state-channels-plus/actions/runs/19128698933/job/54664061954#step:8:59)

lukastanisic99 replied:
when I ran all tests I did see that `revert` (probably out of gas), but when I ran the single test, no error - 


<img width="675" height="253" alt="image" src="https://github.com/user-attachments/assets/178287c7-6b8c-4a97-80bd-01f338a31db5" />


I'll debug more

lukastanisic99 replied:
<img width="977" height="413" alt="image" src="https://github.com/user-attachments/assets/f7dc9507-f682-4d18-8123-8fa62405b294" />


So from the call stack the pipeline has detected a double sign block 

<img width="254" height="346" alt="image" src="https://github.com/user-attachments/assets/f19be684-e6c3-43fa-8295-3f76b348acc0" />


so probably the `junk` posted is not `junk` but a conflicting block - I'll check

lukastanisic99 replied:
@daiagi `postJunkCalldataOnChain` is creating a conflicting block that meets all the requirements for a double sign (same block author, same fork, same height, different block, signed) - that's why the dispute is initiated for double sign (why it fails, that's another story - probably out of gas) - so this is expected behavior 

if you want junk data, you can just change the signature so it's invalid (not signed by the block author) and try with that 

I'll debug to see why the transaction reverted - in general this doubleSign dispute should also be its own test

lukastanisic99 replied:
Debugged typescript - everything working as designed :D will take a look at solidity - this is doing a `multicall` running a fraud proof and chaining a dispute - probably the fraud proof fails :D

lukastanisic99 commented on src/stateManager/ValidationService.ts:
```diff
@@ -164,6 +164,14 @@ export default class ValidationService {
                 existingSignatures
             );
 
+            if (block.onChainTimestamp && !existingBlock.onChainTimestamp) {
```
Does this change fix anything? There could only be 1 on-chain timestamp and currently we don't allow overwriting existing calldata on-chain, so another one wouldn't even be emitted, but in the case that that logic change on-chain we'd expect that the off-chain client also updates the on-chain timestamp (even if set, since the one recored and persisted on-chain is the on relevant)

daiagi replied:
yes, it does change
it fixes a bug: the block that comes in via the "onBlockCalldataPosted" event handler => the onchaintimstamp was not being persisted on the storage block

lukastanisic99 replied:
I assumed that, but don't understand why it's like this - regardless of the presence or absence of an existing block it would be fine to unconditionally update the timestamp in storage if the block (block.hash) is already in storage or am I not getting something?

daiagi replied:
ok so basically you are saying this is sufficient:
`if (block.onChainTimestamp)`

correct?

lukastanisic99 replied:
yes as long as the same block is in storage already e.g. updating by hash `setOnChainTimestamp(block.hash, onChainTimestamp)`

daiagi replied:
fixed

lukastanisic99 commented on test/utils/SyncCoordinator.ts:
```diff
@@ -1,6 +1,11 @@
 import { ForkId, Address } from "@/types/types";
 import { Logger } from "@/utils";
 
+export type WaitForPeersInSyncOptions = {
```
Is this entire file related to tests? If so, why not put it in `test/utils`

daiagi replied:
moved

daiagi committed:
move `syncCoordinator`

daiagi committed:
restore: previousBlockProducerPostedCalldata when calldata was junk

daiagi committed:
simplify  condition for Update the existing block's onChainTimestamp

daiagi committed:
1. harden "disputeCommited" assertion
2. add test case

daiagi committed:
remove skip from test, to reveal bug in logs

daiagi committed:
fix: harden `retrieveSignerAddress` against invalid signatures

daiagi committed:
minor fix: have timeout manger wait for still running tasks (dispose should leave nothing still running)

daiagi committed:
fix: postJunkCalldataOnChain should not trigger double sign dispute

daiagi committed:
add assertions, test passes

daiagi committed:
restore `ECDSA.recover` to `verifyThresholdSigned`

daiagi commented:
@lukastanisic99 
- fixed `postJunkCalldataOnChain`
- fixed `retrieveSignerAddress` in `UtilityFacet` to not revert

good to go now, i think. tests are all passsing

lukastanisic99 commented on src/stateManager/StateManager.ts:
```diff
@@ -1551,6 +1555,11 @@ class StateManager {
         // step 11 - schedule a timeout check for the next participant
+        // Don't schedule timeout tasks if StateManager is disposed
+        if (this.isDisposed) {
+            return;
```
Imo not needed, since `timeoutManager.scheduleTask` considers disposal when scheduling

lukastanisic99 commented on src/stateManager/StateManager.ts:
```diff
@@ -450,6 +450,12 @@ class StateManager {
+
+        // Don't schedule timeout tasks if StateManager is disposed
+        if (this.isDisposed) {
+            return;
```
Yeah better to have this logic in `timeoutManager` so we don't have to think to add this to every place

lukastanisic99 commented on contracts/V1/StateChannelDiamondProxy/UtilityFacet.sol:
```diff
@@ -62,7 +62,12 @@ contract UtilityFacet {
-        return ECDSA.recover(signedHash, signature);
+        // Use tryRecover to handle invalid signatures (recover reverts if signature is invalid)
+        (address recovered, ECDSA.RecoverError error,) = ECDSA.tryRecover(signedHash, signature);
+        if (error != ECDSA.RecoverError.NoError) {
```
so here we just have to do 

```solidity
if (error != ECDSA.RecoverError.NoError) return(recovered, false);
return(recovered, true);
````

For threshold verification we don't need to do this, since the caller should know before calling do they have N correct signatures or not and it's fine for threshold to revert (`recover` calls `tryRecover` and if it's not `NoError` it reverts) - here we want to continue execution in the case of failure so we can potentially slash someone

lukastanisic99 reviewed (changes requested):
Overall fine - just the few things

daiagi committed:
remove `isDisposed` conditionals from StateManager

daiagi committed:
fixup: retrieveSignerAddress

lukastanisic99 reviewed (changes requested):
Seems there are some compiler errors (at least in tests) - let me check

lukastanisic99 commented:
yep, compilation fails locally too

daiagi committed:
fix solidity compile issues

daiagi commented:
oops
fixed
